lib/types/index.d.ts

```diff
@@ -507,175 +547,169 @@ export namespace SourceCode {
+        "Super:exit"?(node: ESTree.Super & NodeParentExtension): void;
+        SwitchCase?(node: ESTree.SwitchCase & NodeParentExtension): void;
+        "SwitchCase:exit"?(node: ESTree.SwitchCase & NodeParentExtension): void;
+        SwitchStatement?(node: ESTree.SwitchStatement & NodeParentExtension): void;
+        "SwitchStatement:exit"?(node: ESTree.SwitchStatement & NodeParentExtension): void;
+        TaggedTemplateExpression?(node: ESTree.TaggedTemplateExpression & NodeParentExtension): void;
+        "TaggedTemplateExpression:exit"?(node: ESTree.TaggedTemplateExpression & NodeParentExtension): void;
+        TemplateElement?(node: ESTree.TemplateElement & NodeParentExtension): void;
+        "TemplateElement:exit"?(node: ESTree.TemplateElement & NodeParentExtension): void;
+        TemplateLiteral?(node: ESTree.TemplateLiteral & NodeParentExtension): void;
+        "TemplateLiteral:exit"?(node: ESTree.TemplateLiteral & NodeParentExtension): void;
+        ThisExpression?(node: ESTree.ThisExpression & NodeParentExtension): void;
+        "ThisExpression:exit"?(node: ESTree.ThisExpression & NodeParentExtension): void;
+        ThrowStatement?(node: ESTree.ThrowStatement & NodeParentExtension): void;
+        "ThrowStatement:exit"?(node: ESTree.ThrowStatement & NodeParentExtension): void;
+        TryStatement?(node: ESTree.TryStatement & NodeParentExtension): void;
+        "TryStatement:exit"?(node: ESTree.TryStatement & NodeParentExtension): void;
+        UnaryExpression?(node: ESTree.UnaryExpression & NodeParentExtension): void;
+        "UnaryExpression:exit"?(node: ESTree.UnaryExpression & NodeParentExtension): void;
+        UpdateExpression?(node: ESTree.UpdateExpression & NodeParentExtension): void;
+        "UpdateExpression:exit"?(node: ESTree.UpdateExpression & NodeParentExtension): void;
+        VariableDeclaration?(node: ESTree.VariableDeclaration & NodeParentExtension): void;
+        "VariableDeclaration:exit"?(node: ESTree.VariableDeclaration & NodeParentExtension): void;
+        VariableDeclarator?(node: ESTree.VariableDeclarator & NodeParentExtension): void;
+        "VariableDeclarator:exit"?(node: ESTree.VariableDeclarator & NodeParentExtension): void;
+        WhileStatement?(node: ESTree.WhileStatement & NodeParentExtension): void;
+        "WhileStatement:exit"?(node: ESTree.WhileStatement & NodeParentExtension): void;
+        WithStatement?(node: ESTree.WithStatement & NodeParentExtension): void;
+        "WithStatement:exit"?(node: ESTree.WithStatement & NodeParentExtension): void;
+        YieldExpression?(node: ESTree.YieldExpression & NodeParentExtension): void;
+        "YieldExpression:exit"?(node: ESTree.YieldExpression & NodeParentExtension): void;
     }
 
     interface NodeParentExtension {
         parent: Node;
     }
     type Node = ESTree.Node & NodeParentExtension;
 
-    interface RuleListener extends NodeListener {
+    type RuleListener = NodeListener & {
```

Comment:
I couldn't get this to validly extend `NodeListener`, so opted for a union type instead, which seemed more appropriate anyway.